#' Read haplotype sharing matrix generated by LIden
#'
#' Create a matrix indicating pedigree haplotype sharing by reading inferred
#' allele sharing table produced by LIden (Wang et al. 2009 BMC Bioinformatics)
#'
#' @param lidenHaploShareTxt
#'      path to inferred_*_allele_sharing.txt in LIden output directory
#' @param lidenHaploSharePlt
#'      path to inferred_*_allele_sharing.plt in LIden output directory
#' @param founderId
#'      string identifier of the founder with respect to the above files
#' @return
#'      integer matrix indicating allele sharing status with the founder. Rows
#'      correspond to chromosomal position, columns to individuals in the
#'      pedigree.
#' @export
readLidenHaploShare <- function(lidenHaploShareTxt, lidenHaploSharePlt,
                                founderId) {

  # Read the haplotype sharing table, truncating the trailing tab on each line
  haploShare <- read.table(pipe(paste(sprintf("zcat -f %s", lidenHaploShareTxt),
                                      "sed -e 's/\t$//'",
                                      sep = " | ")),
                           header = FALSE, sep = "\t", quote = "",
                           row.names = 1, na.strings = "-1",
                           colClasses = "integer", check.names = FALSE,
                           comment.char = "")
  haploShare <- as.matrix(haploShare)
  haploShareFounder <- haploShare[, 1:2]
  haploShare <- haploShare[, 3:ncol(haploShare)]    # Remove founder columns
  # Sanity checks
  if(!all(apply(haploShareFounder, 2, function(x) length(unique(x))) == 1))
    stop("Error: founder haplotype columns are not single valued")
  if(!all(apply(haploShare, 2, function(x) length(unique(x[!is.na(x)]))) < 3))
    stop("Error: haplotype columns have more than 3 codes")

  # The haplotype sharing table specifies haplotype sharing by person (columns)
  # and chromosomal position (rows). There are two problems with this table:
  # (i) columns are not labeled by person ID; and (ii) codes used to
  # distinguish haplotypes are not uniform.
  #
  # To fix these issues, the column labels and haplotype code need to be
  # extracted from the gnuplot file. The "ytics" line in the gnuplot file: (i)
  # lists person ID in order corresponding to the columns in the haplotype
  # sharing table; and (ii) specifies the values used to indicate haplotype
  # sharing in each person.

  # Parse the "ytics" line in the LIden gnuplot file
  haploMeta <- read.table(pipe(paste(sprintf("zcat -f %s", lidenHaploSharePlt),
                                     "sed -n -e 's/^set ytics (\\(.*\\))$/\\1/p'",
                                     "tr ', ' '\\n\\t'",
                                     "sed -e 's/^\\t//'",
                                     sep = " | ")),
                          header = FALSE, sep = "\t",
                          col.names = c("Person", "Code"),
                          colClasses = c("character", "integer"),
                          check.names = FALSE)
  haploMetaFounder <- haploMeta[1:2, ]
  haploMeta <- haploMeta[3:nrow(haploMeta), ]    # Remove founder rows
  personId <- unique(haploMeta$Person)
  # Sanity checks
  if(!all(haploMetaFounder$Person == c("0", "1")))
    stop("Error: founder labels are not 0 and 1")
  if(!all(haploMeta$Person == rep(personId, each = 2)))
    stop("Error: person IDs are not repeated")
  if(!length(personId) == ncol(haploShare))
    stop("Error: number of person IDs does not match haplotype table")

  # 1. Set column (person) names for the haplotype sharing table
  colnames(haploShare) <- personId

  # 2. Unify codes specifying haplotype sharing status to:
  #    0 - haplotype not shared with founder
  #    1 - haplotype shared with homologous chromosome 1 of founder
  #    2 - haplotype shared with homologous chromosome 2 of founder
  #    3 - special code for founder haplotype
  aCodes <- as.integer(by(haploMeta$Code, haploMeta$Person, min))
  bCodes <- as.integer(by(haploMeta$Code, haploMeta$Person, max))
  haploShare[haploShare %in% aCodes] <- as.integer(1)
  haploShare[haploShare %in% bCodes] <- as.integer(2)
  haploShare[is.na(haploShare)]      <- as.integer(0)
  haploShareFounder <- matrix(rep(as.integer(3), nrow(haploShare)), ncol = 1)
  colnames(haploShareFounder) <- founderId
  haploShare <- cbind(haploShare, haploShareFounder)

  return(haploShare)
}
