#' Read EcMap analysis.txt file
#' 
#' Parses the output of EcMap's analysis.txt file into a integer matrix.
#' 
#' EcMap (Yang et al [2008] Eur J Hum Genet) is a program that simulates mapping
#' by allele sharing in a kindred.  EcMap is unversioned and poorly documented. 
#' This function assumes a specific format of the analysis.txt file produced 
#' by the version of EcMap available from the authors ca. Feb 2012, as 
#' exemplified below, 
#' 
#'     The distribution of mutation region
#'             length  0       cM      --      1       cM      Times:  1
#'             length  1       cM      --      2       cM      Times:  1
#'             length  2       cM      --      3       cM      Times:  1
#'     ...
#'             length  49      cM      --      50      cM      Times:  1
#'             length> 50      cM      Times:  0
#'             <blank line>
#'     The distribution of random region
#'             length  0       cM      --      1       cM      Times:  8
#'             length  1       cM      --      2       cM      Times:  15
#'             length  2       cM      --      3       cM      Times:  12
#'     ...
#'             length  27      cM      --      28      cM      Times:  1
#'             length> 28      cM      Times:  0
#' 
#' and returns a two-column matrix of integers as below,
#'  
#'          mutation random
#'     0    1        8
#'     1    1        15
#'     2    1        12
#'     ...
#'     27   ...      1
#'     28   ...      0
#'     ...
#'     49   1        0
#' 
#' Note that values in the last bin (length > n) are assumed to be 0.
#' 
#' @param file
#'     path to an analysis.txt file generated by EcMap
#' @return
#'     integer matrix of counts of IBD region by length in cM (rows) and 
#'     category, causal variant-containing region or region shared by chance.
#' @author Shiv Venkatasubrahmanyam
#' @export
readEcmapAnalysisFile <- function(file) {
    # Locate the headers of each section
    lines <- readLines(file)
    nLines <- length(lines)
    mutationHdrLineNum <- which(lines == "The distribution of mutation region")
    randomHdrLineNum <- which(lines == "The distribution of random region")
    # Read each section separately into a dataframe, skipping the last 
    # "length> n" line
    if(length(mutationHdrLineNum) != 1 || length(randomHdrLineNum) != 1)
        stop("Error locating 'mutation' and/or 'random' header lines")
    mutationHist <- read.table(file, header = FALSE, sep = "", quote = "", 
            row.names = 2, skip = mutationHdrLineNum, 
            nrows = randomHdrLineNum - mutationHdrLineNum - 3, fill = TRUE, 
            blank.lines.skip = TRUE, comment.char = "", 
            stringsAsFactors = FALSE)
    randomHist <- read.table(file, header = FALSE, sep = "", quote = "", 
            row.names = 2, skip = randomHdrLineNum, 
            nrows = nLines - randomHdrLineNum - 1, fill = TRUE, 
            blank.lines.skip = TRUE, comment.char = "", 
            stringsAsFactors = FALSE)
    # Row alignment check
    n <- min(nrow(mutationHist), nrow(randomHist))
    if(any(rownames(mutationHist)[1:n] != rownames(randomHist)[1:n]))
        stop("Error matching rows in 'mutation' and 'random' sections")
    # Construct the result
    nCol <- 2
    nRow <- max(nrow(mutationHist), nrow(randomHist))
    result <- matrix(NA_integer_, nrow = nRow, ncol = nCol)
    colnames(result) <- c("Causal", "Random")
    rownames(result) <- if(nrow(mutationHist) > nrow(randomHist))
            rownames(mutationHist) else rownames(randomHist)
    # Fill the result matrix with the last column from each dataframe
    result[1:nrow(mutationHist), 1] <- mutationHist[, ncol(mutationHist)]
    result[-(1:nrow(mutationHist)), 1] <- 0
    result[1:nrow(randomHist), 2] <- randomHist[, ncol(randomHist)]
    result[-(1:nrow(randomHist)), 2] <- 0
    
    return(result)
}
